[block-serialization-default-parser][fix] Return an empty array early if doc data passed to `parse` is invalid

fullofcaffeine opened the pull request:
## What?

Fixes an edge case where `document` (not to confuse with the DOM's `window.document` object, in this module, `document` is a local variable) is `undefined` and causes a `TypeError`. This happens in some hard-to-reproduce edge cases, and unfortunately I didn't manage to get to the same state in a new test site, but I tested the solution in the test site with the dirty state that caused it.

It happened in the Customizer for a theme that had widget slots. Somehow it got to a state that caused the following error as soon as I opened the `Widgets->Sidebar` and after dismissing the `Welcome to Block Widgtes` welcome message in the widget block editor:

` TypeError: Cannot read properties of undefined (reading 'length')`

In: https://github.com/WordPress/gutenberg/blob/trunk/packages/block-serialization-default-parser/src/index.js#L441

Unfortunately, I don't know how to reproduce the issue consistently, and this fix, while suitable for now, probably does not fix the root issue. More investigation is needed.

## Why?

The bug is causing the widget block editor to completely break.


## How?

By handling `undefined` (js) /`null` (PHP) and returning early from `parse` with an empty array, and avoiding other functions to be called that wouldn't expect `doc` / `$document` to be nullish, which would cause an exception later on.

<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## Testing Instructions

Apart from the ones I added in the `What` section, I don't have complete steps to reproduce now, unfortunately.

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

fullofcaffeine committed:
Add safe nav op for accessing `length` for `document` in `addFreeForm()` as in some scenarios, `document` can be `undefined`

In some hard-to-reproduce (possibly edgy) scenarios, the `document` variable in the Index module for the `block-serialiazation-parser` package can be `undefined`. This is causing the following error in some WP instances/themes that support widgets + block widgets:

```
TypeError: Cannot read properties of undefined (reading 'length')
```

In: https://github.com/WordPress/gutenberg/blob/trunk/packages/block-serialization-default-parser/src/index.js#L441

Unfortunately, I don't know how to consistently reproduce the issue, and this fix, while suitable for now, probably does not fix the root issue. More investigation is needed.

fullofcaffeine commented on packages/block-serialization-default-parser/src/index.js:
```diff
@@ -320,10 +320,10 @@ function proceed() {
 				//  - assume an implicit opener
 				//  - assume _this_ is the opener
 				// - give up and close out the document.
+
```
Ops, I'll remove this, added by accident.

dmsnell reviewed (commented):
Noted in Slack, I'd rather see us validate the input document inside `parse` than cover up the problem deep in `addFreeform`

That is, the only possible way `document` is undefined here is if we were passed `undefined` as the input HTML, so we can bail as soon as we start if that happens.

The only question is whether to propagate some kind of error by returning `null` or return an empty set of blocks, which won't create an error but will likely result in data corruption, because code expects blocks and gets none.

That being said, if we check `'string' === typeof doc` inside `parse()` then to some extent it's not the parser's problem if given invalid input. Anything that's not a string containing blocks, when parsed, contains no blocks, so `[]` is one valid way to interpret or expect the results of calling `parse(undefined)` or `parse()` or `parse(5)` or `parse([])`

fullofcaffeine committed:
Return an empty array early if doc data passed to parse is nullish

It could happen in some edge casesthat the value for `doc` (JS) or `$document` (PHP) is nullish (`undefined` or `null`). If that's the case, we return an empty array early to avoid other functions that don't expect a nullish doc to run and breaking the execution with an exception.

dmsnell commented on packages/block-serialization-default-parser/parser.php:
```diff
@@ -225,6 +225,10 @@ class WP_Block_Parser {
 	 * @return array[]
 	 */
 	public function parse( $document ) {
+		if ( gettype( $document ) !== 'string' ) {
```
```php
if ( ! is_string( $document ) )
```

dmsnell commented:
@fullofcaffeine we could update the PR summary and squash the commits with a new summary and description. this isn't really about `nullish` but rather something more like "parser: return empty array when given invalid input" and "if given any value that isn't a string then there can be no block content to parser, therefore return an empty list"

fullofcaffeine commented:
Closed in favor of https://github.com/WordPress/gutenberg/pull/46487.
